### ISSUE TITLE
[FEA] Priority support in cuda_stream_pool

### ISSUE THREAD

**Issue by seunghwak** [rapidsai/rmm#939]
**Is your feature request related to a problem? Please describe.**
In multi-node multi-GPU execution, running communication-bound kernel with higher-priority helps in overlapping communication with computation.

Say kernel A has little computing and bigger inter-GPU communication. Kernel B has way more computing but requires only little inter-GPU communication.

If we run kernel A on a higher priority stream, kernel A will finish computation as soon as possible and start communication while kernel B is still working on local computing; this effectively overlaps communication with computation.

rmm's cuda_stream_pool currently does not support stream priorities and I cannot implement this strategy using rmm::cuda_stream_pool.

**Describe the solution you'd like**
I should be able to get a higher priority stream using (a variant of) `get_stream` or should be able to create a stream pool with a priority. A single pool may manage streams with different priorities (based on index ranges) or application may maintain multiple pools (one pool per priority).

**Comment by harrism** [comment]
@seunghwak have you prototyped this with raw CUDA streams to verify that you get the benefits you expect?

**Comment by seunghwak** [comment]
> @seunghwak have you prototyped this with raw CUDA streams to verify that you get the benefits you expect?

Yes, this was used to produced the PageRank demo results for GTC 2020. I'm working on bringing this optimization to cugraph.

**Comment by harrism** [comment]
Can you share some specifics (speedups) here to help motivate?

**Comment by seunghwak** [comment]
Thanks, and I will collect numbers in some days (after fixing another issue I am currently working on) and post here.

**Comment by seunghwak** [comment]
So, this varies on the input graphs and the target system (relative speed of computing vs network bandwidth), but for Graph 500 style input graphs and 16 GPUs, using priorities cut 15 % of the execution time (relative to multi-stream execution using the same priority) for the most expensive kernel in PageRank and 12% of the total PageRank execution time.

Currently, I have tested this for only one part of the cuGraph code, but there are many other places to apply the same optimization, so this can have similar impacts on a range of graph algorithms.

**Comment by harrism** [comment]
@seunghwak not working on this yet, but thinking about it. How many priorities do you use in your example?

**Comment by bdice** [comment]
There are two proposed designs here. One is to allow a stream pool to return streams with different priorities. The other is to allow stream pools to be constructed with a given priority for all streams.

Only the second design (constructing a pool with a given priority for all streams) makes sense to me. The purpose of a pool is to reduce the cost of creating/destroying streams by reusing them. Because priorities can be arbitrary, there's not a good way to reuse streams unless they were already created with the desired priority. See [cudaStreamCreateWithPriority docs](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1ge2be9e9858849bf62ba4a8b66d1c3540).

If we move forward with this, the action item I would recommend is to add a constructor argument for `rmm::cuda_stream_pool` that accepts a priority and uses it for all streams in the pool. Applications would then be responsible for using multiple pools based on the desired priority.

**Comment by harrism** [comment]
We could have a pool that returns streams with a priority.  When you request a stream from the pool, it returns an existing stream with that priority if it exists, or if not it returns a new stream with that priority.

The implementation could use a collection of fixed-priority pools under the hood.

**Comment by vyasr** [comment]
I agree with Bradley that the first use case (allow a stream pool to return streams with different priorities) doesn't really make sense to me. @seunghwak could you elaborate on how you imagined that version working? I'm referring to
> A single pool may manage streams with different priorities (based on index ranges) 

At that point what would be the benefit of the pool over creating streams directly?

Perhaps what is really desired at that level is lifetime management of the stream, in which case what we could do (and probably what we ought to do in any case) is to amend `rmm::cuda_stream`'s constructor to accept a priority so that streams can be constructed with priorities. Then the pool could simply construct streams of a desired priority (most likely, as discussed above, all streams in a pool would have the same priority, the pool's priority).
